packages/server-wallet/src/protocols/__test__/process-ledger-queue.test.ts

```diff
@@ -0,0 +1,380 @@
+import _ from 'lodash';
+import {
+  AllocationItem,
+  BN,
+  makeDestination,
+  Participant,
+  SimpleAllocation,
+  simpleEthAllocation as coreSimpleEthAllocation,
+} from '@statechannels/wallet-core';
+
+import {channel} from '../../models/__test__/fixtures/channel';
+import {alice, bob} from '../../wallet/__test__/fixtures/participants';
+import {alice as aliceSW, bob as bobSW} from '../../wallet/__test__/fixtures/signing-wallets';
+import {stateSignedBy} from '../../wallet/__test__/fixtures/states';
+import {protocol, ProtocolState} from '../process-ledger-queue';
+import {Fixture} from '../../wallet/__test__/fixtures/utils';
+import {ChannelStateWithSupported} from '../state';
+
+// TEST HELPERS
+// There are many test cases in this file. These helpers make the tests cases more readable.
+
+type PreAllocationItem = [string | Fixture<Participant>, number];
+
+function allocationItem(preAllocationItem: PreAllocationItem): AllocationItem {
+  const [destinationOrFixture, amount] = preAllocationItem;
+  const destination =
+    typeof destinationOrFixture === 'string'
+      ? makeDestination(destinationOrFixture)
+      : destinationOrFixture().destination;
+  return {destination, amount: BN.from(amount)};
+}
+
+function simpleEthAllocation(...items: PreAllocationItem[]): SimpleAllocation {
+  return coreSimpleEthAllocation(items.map(allocationItem));
+}
+
+let i = 0;
+const prefundChannelWithAllocations = (...allocations: PreAllocationItem[]) =>
+  channel({
+    channelNonce: i++, // Unique channel id each time this fixture is used
+    fundingStrategy: 'Ledger',
+    vars: [
+      stateSignedBy([aliceSW(), bobSW()])({
+        turnNum: 1,
+        outcome: simpleEthAllocation(...allocations),
+      }),
+    ],
+  });
+
+const ledgerChannelWithAllocations = (...allocations: PreAllocationItem[]) =>
+  channel({
+    vars: [
+      stateSignedBy([aliceSW(), bobSW()])({
+        turnNum: 3,
+        outcome: simpleEthAllocation(...allocations),
+      }),
+    ],
+  }).protocolState as ChannelStateWithSupported;
+// END TEST HELPERS
+
+const defaultLedgerChannel = ledgerChannelWithAllocations([alice, 10], [bob, 10]);
+
+// const unfundedLedgerChannel = ledgerChannelWithAllocations([]);
+
+const processLedgerQueueProtocolState = (args: Partial<ProtocolState> = {}) => ({
+  fundingChannel: defaultLedgerChannel,
+
+  channelsRequestingFunds: [],
+  channelsReturningFunds: [],
+
+  theirLedgerProposal: undefined,
+  theirLedgerProposalNonce: 0,
+
+  myLedgerProposal: undefined,
+  myLedgerProposalNonce: 0,
+
+  ...args,
+});
+
+describe('marking ledger requests as complete', () => {
```

Comment:
I didn't read these test cases carefully, but I skimmed them, and they generally make sense to me.

I have the feeling that there are edge cases that can lead to deadlocks. (But we're aware of deadlocks that are fixed by this change.) I'd like to explore some strategies for detecting potential issues with a long-running randomized test.